src/app/clusters/thermostat-server/thermostat.cpp

```diff
@@ -45,115 +45,182 @@ void emberAfThermostatClusterServerInitCallback(void)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterGetWeeklyScheduleCallback(app::CommandHandler * commandObj, uint8_t daysToReturn, uint8_t modeToReturn)
+bool emberAfThermostatClusterGetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t daysToReturn, uint8_t modeToReturn)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterSetWeeklyScheduleCallback(app::CommandHandler * commandObj, uint8_t numberOfTransitionsForSequence,
-                                                       uint8_t daysOfWeekForSequence, uint8_t modeForSequence, uint8_t * payload)
+bool emberAfThermostatClusterSetWeeklyScheduleCallback(EndpointId aEndpointId, chip::app::CommandHandler * commandObj,
+                                                       uint8_t numberOfTransitionsForSequence, uint8_t daysOfWeekForSequence,
+                                                       uint8_t modeForSequence, uint8_t * payload)
 {
     // TODO
     return false;
 }
 
-bool emberAfThermostatClusterSetpointRaiseLowerCallback(app::CommandHandler * commandObj, uint8_t mode, int8_t amount)
+int32_t EnforceHeatingSetpointLimits(int32_t HeatingSetpoint, EndpointId endpoint)
+{
+    // Optional Mfg supplied limits
+    int16_t AbsMinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t AbsMaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Optional User supplied limits
+    int16_t MinHeatSetpointLimit = 0x02bc; // 7C (44.5 F) is the default
+    int16_t MaxHeatSetpointLimit = 0x0bb8; // 30C (86 F) is the default
+
+    // Attempt to read the setpoint limits
+    // Absmin/max are manufacturer limits
+    // min/max are user imposed min/max
+
+    // Note that the limits are initialized above per the spec limits
+    // if they are not present emberAfReadAttribute() will not update the value so the defaults are used
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMinHeatSetpointLimit, sizeof(AbsMinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_ABS_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &AbsMaxHeatSetpointLimit, sizeof(AbsMaxHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MIN_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MinHeatSetpointLimit, sizeof(MinHeatSetpointLimit), NULL);
+
+    emberAfReadAttribute(endpoint, ZCL_THERMOSTAT_CLUSTER_ID, ZCL_MAX_HEAT_SETPOINT_LIMIT_ATTRIBUTE_ID, CLUSTER_MASK_SERVER,
+                         (uint8_t *) &MaxHeatSetpointLimit, sizeof(MaxHeatSetpointLimit), NULL);
+
+    // Make sure the user imposed limits are within the manufacture imposed limits
```

Comment:
```suggestion
    // Make sure the user imposed limits are within the manufacturer imposed limits
```
I guess this does not happen at attribute write time, per spec?

Also, where in the spec is it specified that enforcement on setpoints takes the Abs* values into account at all?  Does this need a spec fix to actually specify the behavior here?

Reply:
It was not clear to me that limits were enforced at attribute write time.  

The spec is unclear regarding Abs* limits.  Only a reference I could find is the if Min*/Max* are not present the Abs* is used.  All these limit attributes are of independently optional.  Perhaps a spec change is in order.  The DM team accepted the existing zcl spec.  
My goal here is to implement what I see as the intent here.  A manufacturer could/should set these Abs* values to reflect the capabilities of their device.  The min and max limits are there for user use to be more restrictive than the manufacturer imposed limits.

Reply:
I did not mean to imply that the spec says to enforce the limits at attribute write time.  The spec generally doesn't define how these things are supposed to work.  My point is that it _should_ define it.

> The spec is unclear regarding Abs* limits.

Yes.  I am saying we need spec issues tracking that, make the spec clear, then make sure this code follows the spec (probably in a followup; I don't see a need to block this PR on the spec work).

Can you file the relevant spec issues and make sure they get on the HVAC group's agenda?

> The DM team accepted the existing zcl spec.

Yes, I'm aware of that.  I'm just saying the spec does not define things sufficiently clearly that we can create interoperable implementations based on it, as things stand.

> My goal here is to implement what I see as the intent here.

I understand the intent, sure.  And again, I am not saying this PR should block on the spec.  I'm just saying we need to fix the spec.